### PR TITLE
[SPARK-14124] [SQL] Implement Database-related DDL Commands

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

This PR is to implement the following four Database-related DDL commands:
- `CREATE DATABASE|SCHEMA [IF NOT EXISTS] database_name`
- `DROP DATABASE [IF EXISTS] database_name [RESTRICT|CASCADE]`
- `DESCRIBE DATABASE [EXTENDED] db_name`
- `ALTER (DATABASE|SCHEMA) database_name SET DBPROPERTIES (property_name=property_value, ...)`

Another PR will be submitted to handle the unsupported commands. In the Database-related DDL commands, we will issue an error exception for `ALTER (DATABASE|SCHEMA) database_name SET OWNER [USER|ROLE] user_or_role`.

cc @yhuai @andrewor14 @rxin Could you review the changes? Is it in the right direction? Thanks!
#### How was this patch tested?

Added a few test cases in `command/DDLSuite.scala` for testing DDL command execution in `SQLContext`. Since `HiveContext` also shares the same implementation, the existing test cases in `\hive` also verifies the correctness of these commands.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -104,6 +104,13 @@ class SessionCatalog(externalCatalog: ExternalCatalog, conf: CatalystConf) {
     currentDb = db
   }
 
+  // todo: what is the default path in SessionCatalog?
+  def getDefaultPath: String = ""
+
+  def getDefaultDBExtension: String = ".db"
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: what is this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -39,7 +39,7 @@ abstract class ExternalCatalog {
 
   protected def requireDbExists(db: String): Unit = {
     if (!databaseExists(db)) {
-      throw new AnalysisException(s"Database $db does not exist")
+      throw new AnalysisException(s"Database '$db' does not exist")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: maybe use backtick?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: That is from Hive. http://www.reedbushey.com/99Programming%20Hive.pdf

> The database directory is created under a top-level directory specified by the property hive.metastore.warehouse.dir, which we discussed in “Local Mode Configuration” on page 24 and “Distributed and Pseudodistributed Mode Configura- tion” on page 26. Assuming you are using the default value for this property, /user/hive/ warehouse, when the financials database is created, Hive will create the directory /user/ hive/warehouse/financials.db. Note the .db extension.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Ok. Sure. We need to change all the places.

Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: let's change it in a separate patch then. There are a few other places like this

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+   */
+  private def withDatabase(dbNames: String*)(f: => Unit): Unit = {
+    try f finally {
+      dbNames.foreach { name =>
+        sqlContext.sql(s"DROP DATABASE IF EXISTS $name CASCADE")
+      }
+    }
+  }
+
+  test("Create/Drop/Alter/Describe Database - basic") {
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: can you split these into separate tests?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Sure, will do it in a separate PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Sure, will do it. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 1 above)
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: either way I don't think we want this; just embed it in `getDefaultDBPath`. Otherwise we'll end up with too many random methods

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 4]
```diff
@@ -104,6 +104,13 @@ class SessionCatalog(externalCatalog: ExternalCatalog, conf: CatalystConf) {
     currentDb = db
   }
 
+  // todo: what is the default path in SessionCatalog?
+  def getDefaultPath: String = ""
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: this is only used for the database right? I'd remove this and introduce only `getDefaultDBPath`. For `SessionCatalog` I would just use the temp dir.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 5]
```diff
@@ -318,6 +320,138 @@ case class DescribeCommand(
...
+ * it is not empty. The inclusive tables must be dropped at first.
+ *
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    DROP DATABASE [IF EXISTS] database_name [RESTRICT|CASCADE];
+ * }}}
+ */
+case class DropDatabase(
+    databaseName: String,
+    ignoreIfNotExists: Boolean,
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: let's call this `ifExists` to be consistent

**Comment by andrewor14** [comment]
Looks great.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 6]
```diff
@@ -318,6 +320,138 @@ case class DescribeCommand(
...
+ * A command for users to create a new database.
+ *
+ * It will issue an error message when the database with the same name already exists,
+ * unless 'ifNotExists' is true.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    CREATE DATABASE|SCHEMA [IF NOT EXISTS] database_name
+ * }}}
+ */
+case class CreateDatabase(
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: also, why did you move them to this file? These are DDLs so they belong to `ddl.scala`

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Agreed. We should keep them in `ddl.scala`.

**Comment by rxin** [comment]
I just merged https://github.com/apache/spark/pull/12015

Can you update to use the new ANTLR4 parser instead? We are going to remove the ANTLR3 one in the next day or two. Thanks.

**Comment by gatorsmile** [comment]
Sure, will do the changes. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Sure, will move them back. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Sure, will do. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: ok, will do.

**Comment by andrewor14** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala [hunk 7]
```diff
@@ -97,7 +97,8 @@ private[sql] class SparkQl(conf: ParserConf = SimpleParserConf()) extends Cataly
 
       // CREATE DATABASE [IF NOT EXISTS] database_name [COMMENT database_comment]
       // [LOCATION path] [WITH DBPROPERTIES (key1=val1, key2=val2, ...)];
-      case Token("TOK_CREATEDATABASE", Token(databaseName, Nil) :: args) =>
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkQl.scala]: changes in this file aren't really necessary since the file will be deleted anyway, but for now it's OK to keep

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 8]
```diff
@@ -45,46 +45,135 @@ abstract class NativeDDLCommand(val sql: String) extends RunnableCommand {
...
     databaseName: String,
-    extended: Boolean)(sql: String)
-  extends NativeDDLCommand(sql) with Logging
+    extended: Boolean)
+  extends RunnableCommand {
+
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val dbMetadata: CatalogDatabase = sqlContext.sessionState.catalog.getDatabase(databaseName)
+    val result =
+      Row("Database Name", dbMetadata.name) ::
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: this can probably just be `Name`. The user ran `DESCRIBE DATABASE` so it's pretty obvious

**Comment by andrewor14** [comment]
LGTM will merge this once tests pass.

**Comment by andrewor14** [comment]
Oh good timing. Merging into master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 8 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I'll change this myself

**Comment by gatorsmile** [comment]
Thank you for your reviews!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 9]
```diff
@@ -45,46 +45,135 @@ abstract class NativeDDLCommand(val sql: String) extends RunnableCommand {
...
+    val catalog = sqlContext.sessionState.catalog
+    catalog.createDatabase(
+      CatalogDatabase(
+        databaseName,
+        comment.getOrElse(""),
+        path.getOrElse(catalog.getDefaultDBPath(databaseName)),
+        props),
+      ifNotExists)
+    Seq.empty[Row]
+  }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Do we need to create the underlying dir in this command?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: True. Only did it in the default DB path. I forgot to do it in the regular case : ( Let me submit a follow-up PR for it. Thanks for catching it!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I tried it in `spark-sql`. If the directory is not created, Hive will do it for us. I am wondering if we still should create directory in Spark?

However, this PR has an issue when users specify the location in the `Create Database` command. The generated path should be `path/databaseName.db` instead of `path`. Will fix it soon.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Yea. Let's create the directory if it is not created.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I see. Will do it in https://github.com/apache/spark/pull/12081. Thanks!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: @yhuai I did try it. Actually, the code is done... However, if we create a directory before issuing Hive client API `createDatabase`, we will get the following error message from Hive:
```
Error in query: org.apache.hadoop.hive.metastore.api.AlreadyExistsException: Database db3 already exists;
```

Just feel free to let me know what I should do next. Thanks!
